include/dali/core/tensor_layout.h

```diff
@@ -0,0 +1,475 @@
+
+  DALI_HOST_DEV
+  constexpr iterator begin() noexcept               { return data_; }
+  DALI_HOST_DEV
+  constexpr iterator end() noexcept                 { return data_ + ndim(); }
+  DALI_HOST_DEV
+  constexpr auto begin() const noexcept             { return cbegin(); }
+  DALI_HOST_DEV
+  constexpr auto end() const noexcept               { return cend(); }
+  DALI_HOST_DEV
+  constexpr const_iterator cbegin() const noexcept  { return data_; }
+  DALI_HOST_DEV
+  constexpr const_iterator cend() const noexcept    { return data_ + ndim(); }
+
+  DALI_HOST_DEV
+  TensorLayout sample_layout() const {
+    if (empty())
+      return {};
+#ifdef __CUDA_ARCH__
+    assert(data_[0] == 'N');
+#else
+    if (find('N') != 0)
+      throw std::logic_error("This is not a multi-sample layout: \"" + str() + "\"");
+#endif
+    return sub(1);
+  }
+
+  DALI_HOST_DEV
+  constexpr TensorLayout sub(int start_dim, int n = -1) const noexcept {
+    assert(start_dim >= 0 && start_dim <= ndim());
+    if (n < 0)
+      n = ndim() - start_dim;
+    assert(start_dim + n <= ndim());
+    return TensorLayout(begin() + start_dim, n);
+  }
+
+  DALI_HOST_DEV
+  constexpr TensorLayout first(int n) const noexcept {
+    assert(n <= ndim());
+    return TensorLayout(begin(), n);
+  }
+
+  DALI_HOST_DEV
+  constexpr TensorLayout last(int n) const noexcept {
+    assert(n <= ndim());
+    return TensorLayout(end() - n, n);
+  }
+
+  static constexpr int max_ndim = 15;
+
+ private:
+  /** @brief Stores the dimension descriptions as a null-terminated string
+   *
+   * Last character in the string is aliased with max_ndim - length.
+   * If length == max_ndim, then 0 is stored and aliased as '\0' terminator,
+   * allowing 1 extra character to be stored.
+   */
+  char data_[max_ndim + 1] = { 0 };
+  DALI_HOST_DEV
+  void set_size(int n) {
+    assert(n >= 0 && n <= max_ndim);
+    data_[max_ndim] = max_ndim - n;
+  }
+  DALI_HOST_DEV
+  friend constexpr TensorLayout operator+(const TensorLayout &a, const TensorLayout &b);
+};
+
+static_assert(sizeof(TensorLayout) == 16, "Tensor layout size should be exactly 16B");
+
+DALI_HOST_DEV
+constexpr TensorLayout operator+(const TensorLayout &a, const TensorLayout &b) {
+  assert(a.size() + b.size() < TensorLayout::max_ndim);
+  TensorLayout result = a;
+  int i = 0, j = 0;
+  for (i = result.ndim(), j = 0; i < TensorLayout::max_ndim && j < b.ndim(); i++, j++)
+    result[i] = b[j];
+  result[i] = 0;
+
+  /* Cannot use
+   *   result.set_size(i);
+   * because it's not constexpr
+   */
+  result.data_[TensorLayout::max_ndim] = TensorLayout::max_ndim - i;
+  return result;
+}
+
+
+DALI_HOST_DEV
+constexpr TensorLayout TensorLayout::skip(char dim_name) const noexcept {
+  int i = find(dim_name);
+  if (i < 0)
+    return *this;
+  return first(i) + sub(i+1);
+}
+
+#define DEFINE_TENSOR_LAYOUT_COMPARISON(op)                             \
+inline bool operator op(const TensorLayout &tl, const std::string &s) { \
```

Comment:
maybe personal preference, but I'd name the argument as something that screams "I am a macro"
e.g. `OP` or `OPERATOR`